helix-view/src/editor.rs

```diff
@@ -799,6 +799,26 @@ pub struct Breakpoint {
 
 use futures_util::stream::{Flatten, Once};
 
+/// Threadsave (Sync + Send) iteration over document text (and path)
+/// This is necessary because the editor and the document contain
+/// interior mutability which makes it impossible to send them across threads.
+/// However the document text itself (ropes) is perfectly fine to send across
+/// threads
+#[derive(Clone)]
+pub struct SharedDocuments<'a>(btree_map::Values<'a, DocumentId, Document>);
+
+unsafe impl Sync for SharedDocuments<'_> {}
+unsafe impl Send for SharedDocuments<'_> {}
+
+impl<'a> Iterator for SharedDocuments<'a> {
+    type Item = (Option<&'a PathBuf>, &'a Rope);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let doc = self.0.next()?;
+        Some((doc.path(), doc.text()))
+    }
+}
```

Comment:
Could we avoid this custom type by `documents.values().map(..).collect()`, then returning the Vec? It would mean allocating an intermediary vec but it's unlikely the user will have enough documents open for it to make a difference

Reply:
you are totally right. I think back then I sort of started out trying that but with the full document because that doesn't work thanks to the cell. Then I implemented this iterator (with manual sync/send) and later only shared path and rope to make the unsave code sound but somehow missed the obviour thing that I could have just collect the rope/path pairs to begin with. 

I have removed the custom type, sorry about that!